### ISSUE TITLE
Unable to connect to application

### ISSUE THREAD

**Issue by LLukens** [jcrist/skein#45]
We are having some challenges getting dask-distributed running on skein/dask-yarn....following error shows up when running on edge.   Working on large cluster,  Cloudera (5.14 w/ Kerberos).   Happy to help support testing, debug, etc. on this project going forward.

18/07/26 17:55:08 INFO skein.Daemon: Submitting application...
18/07/26 17:55:08 INFO impl.YarnClientImpl: Submitted application application_1531262409700_2872
18/07/26 17:55:14 INFO skein.Daemon: Notifying that application_1531262409700_2872 has started. 1 callbacks registered.
18/07/26 17:55:14 INFO skein.Daemon: Removing callbacks for application_1531262409700_2872
18/07/26 17:55:26 INFO impl.YarnClientImpl: Killed application application_1531262409700_2872
Traceback (most recent call last):
  File "daskyarn_gist.py", line 9, in <module>
    name='daskyarn')
  File "/home/reporting/miniconda3/envs/daskyarn/lib/python3.6/site-packages/dask_yarn/core.py", line 175, in __init__
    self._start_cluster(spec, skein_client)
  File "/home/reporting/miniconda3/envs/daskyarn/lib/python3.6/site-packages/dask_yarn/core.py", line 201, in _start_cluster
    scheduler_address = application_client.kv.wait('dask.scheduler')
  File "/home/reporting/miniconda3/envs/daskyarn/lib/python3.6/site-packages/skein/core.py", line 553, in wait
    return self._get(key=key, wait=True)
  File "/home/reporting/miniconda3/envs/daskyarn/lib/python3.6/site-packages/skein/core.py", line 548, in _get
    resp = self._client._call('keyvalueGetKey', req)
  File "/home/reporting/miniconda3/envs/daskyarn/lib/python3.6/site-packages/skein/core.py", line 266, in _call
    raise ConnectionError("Unable to connect to %s" % self._server_name)
skein.exceptions.ConnectionError: Unable to connect to application
18/07/26 17:55:26 INFO skein.Daemon: Starting process disconnected, shutting down
18/07/26 17:55:26 INFO skein.Daemon: Shutting down gRPC server
18/07/26 17:55:26 INFO skein.Daemon: gRPC server shut down

Thanks!

**Comment by superbobry** [comment]
Can you attach the app master logs, please?

**Comment by LLukens** [comment]
Attached....replaced user, IP with X for security purposes.   Do greatly appreciate the help.  Thank you!

[appmaster.log](https://github.com/jcrist/skein/files/2236231/appmaster.log)

**Comment by superbobry** [comment]
Hmm, looks like the scheduler failed. Could you additionally include the corresponding log from container_1531262409700_2997_01_000002?

**Comment by jcrist** [comment]
Apologies for letting this sit @LLukens. Any updates? From looking at the log you provided, it looks like the issue might be lack of support for HA mode in Skein (https://wiki.apache.org/hadoop/FAQ#What_does_the_message_.22Operation_category_READ.2FWRITE_is_not_supported_in_state_standby.22_mean.3F).

This could potentially be fixed by something like #58 along with automatically detecting high-availability mode and requesting delegation tokens for every namenode by default.

**Comment by LLukens** [comment]
Sorry for the delay getting back.  Last week was travel.  The scheduler not starting was the problem, although was not able to locate logs to pinpoint exactly why.  I had difficulty getting the scheduler container logs to retain on Hadoop, even after changing yarn log retention settings.   Dask workers okay, my workaround right now is just calling skein passing required Kerberos credentials along with private scheduler IP to start workers. Then initiating separate dask-scheduler job.  Working great but do want to use dask-yarn as its all-in-one.  I’ll give it another try this week.   Good news is Dask is showing better performance on dataframes, on my use case over Spark w/ less head scratching java errors...win-win!

Sent from my iPhone

On Aug 14, 2018, at 2:52 PM, Jim Crist <notifications@github.com<mailto:notifications@github.com>> wrote:

CAUTION: This email originated from outside of the organization. Do not click links or open attachments unless you recognize the sender and know the content is safe.


Apologies for letting this sit @LLukens<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FLLukens&data=02%7C01%7Clance.p.lukens%40sprint.com%7Cd12e1396548b4327770d08d602303c4a%7C4f8bc0acbd784bf5b55f1b31301d9adf%7C0%7C0%7C636698803522592888&sdata=Cxa11uwvQwxtrLe2%2FVe8%2FUhQtyGYOl20fn%2BaYr8nLUc%3D&reserved=0>. Any updates? From looking at the log you provided, it looks like the issue might be lack of support for HA mode in Skein (https://wiki.apache.org/hadoop/FAQ#What_does_the_message_.22Operation_category_READ.2FWRITE_is_not_supported_in_state_standby.22_mean.3F<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fwiki.apache.org%2Fhadoop%2FFAQ%23What_does_the_message_.22Operation_category_READ.2FWRITE_is_not_supported_in_state_standby.22_mean.3F&data=02%7C01%7Clance.p.lukens%40sprint.com%7Cd12e1396548b4327770d08d602303c4a%7C4f8bc0acbd784bf5b55f1b31301d9adf%7C0%7C0%7C636698803522602893&sdata=UqwxZfGc9nXaKGgzPEU%2B2FbEcRuPyP8TtaO5nVxIjf4%3D&reserved=0>).

This could potentially be fixed by something like #58<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fjcrist%2Fskein%2Fpull%2F58&data=02%7C01%7Clance.p.lukens%40sprint.com%7Cd12e1396548b4327770d08d602303c4a%7C4f8bc0acbd784bf5b55f1b31301d9adf%7C0%7C0%7C636698803522602893&sdata=8PD3hsrsnVMWIih73Wr4SkHeWEuADpMeR0gy7SmJF9g%3D&reserved=0> along with automatically detecting high-availability mode and requesting delegation tokens for every namenode by default.

**Comment by LLukens** [comment]
Think to have found the issue....there's no IP address provided from the scheduler start.    core.py  addr = uri_from_host_port('', None, 0) returns tcp:// missing the IP and port.  So, workers are not getting passed the scheduler ip:port to connect into.     I need to dig into but maybe it's that we have Ethernet name "bond0" as default on yarn workers and not picking up....instead of common "eth0"?   If you have any ideas or have come across this issue before please let me know.

**Comment by jcrist** [comment]
>  returns tcp:// missing the IP and port. 

That line actually doesn't do anything, the `scheduler.start(addr)` line could be replaced with `scheduler.start('tcp://')` as equivalent. This looks to be a bug in dask distributed's address parsing logic that just happens to work out in the end.

The address that's registered with the skein kv store is pulled off the scheduler object after it's started, and has an actual value: https://github.com/dask/dask-yarn/blob/master/dask_yarn/cli/dask_yarn_scheduler.py#L48

> my workaround right now is just calling skein passing required Kerberos credentials along with private scheduler IP to start workers. Then initiating separate dask-scheduler job.

I'm a bit confused here. You're saying starting two separate jobs (one for scheduler, one for workers) works? I assume you have a manually specified scheduler address in this case? If that's the case then:

- You can provide a single skein specification to launch everything in a single YARN job using `YarnCluster.from_specification` or the [configuration](https://dask-yarn.readthedocs.io/en/latest/#configuration). There's no reason to have two jobs here.

- That would indicate it probably is a bug in the random port/address selection code. Logs would be quite helpful here, I'm not sure how else to help diagnose this.

**Comment by LLukens** [comment]
Fixed!   My colleague reminded me of recent commits over the past week.   I just pulled down the latest and solved my issue.   I do see now uri_from_host_port wasn't the cause, kv.wait you highlighted was it as now see a IP address returned on that step.    Wasn't able to capture any logs in the 02 yarn scheduler container that made this a challenge.  Correct, my temp workaround was to run a dask-scheduler outside of yarn on a single host, separate job.  Now having dask-yarn working no longer need as can do all-in-one. Thanks again for all the help and support!
